Implement the XSETTINGS protocol

Stebalien opened the pull request:
Users can use this to configure system-wide themes, icons, fonts, etc.

```elisp
(setopt exwm-xsettings-theme '("Adwaita" . "Adwaita-dark")
        exwm-xsettings-icon-theme '("Papirus-Light" . "Papirus-Dark")
        exwm-xsettings '(("Gtk/FontName" . "IBM Plex Sans 9")
                         ("Gtk/DocumentFontName" . "IBM Plex Sans 9")
                         ("Gtk/MonospaceFontName" . "JetBrains Mono 9")))
```

fixes #876

NOTE: If enabled, this module will conflict with any other XSETTINGS implementation and will refuse to start.

---

* exwm-xsettings.el: Implement the XSETTINGS protocol (fixes #876)

Stebalien committed:
Implement the XSETTINGS protocol

fixes #876

Stebalien commented on exwm-xsettings.el:
```diff
@@ -0,0 +1,343 @@
+                       :destination exwm--root
+                       :event-mask xcb:EventMask:StructureNotify
+                       :event (xcb:marshal
+                               ;; TODO: new event? Do I need a new field?
```
This is one of the things I'm not sure about. I'm "borrowing" the systemtray client message struct because it's identical, but maybe I should define a new one?

medranocalvo replied:
Yes, please.  Copypaste it here with a new name.  They are different messages although at this time they both look same.

medranocalvo replied:
It seems that both are selection announcement messages (https://x.org/releases/X11R7.6/doc/xorg-docs/specs/ICCCM/icccm.html#manager_selections).  I think it is better to have two different messages, one for each manager.

Stebalien replied:
Fixed.

Stebalien commented on exwm-xsettings.el:
```diff
@@ -0,0 +1,343 @@
+                 (cons (string :tag "Light Icon Theme")
+                       (string :tag "Dark Icon Theme"))))
+
+(defvar exwm-xsettings--serial 0)
+
+(defconst xcb:xsettings:-Type:Integer 0)
+(defconst xcb:xsettings:-Type:String 1)
+(defconst xcb:xsettings:-Type:Color 2)
+
+(defclass xcb:xsettings:-Settings
```
Should I move some of this to xelb?

medranocalvo replied:
Hmmm, good question. (I'm thinking out loud here.)  Most (all?) definitions like this in XELB are generated from the XML protocol definitions (xcb-xproto).  The XSettings spec is not part of xcb-xproto and there does not seem to exist a machine-readable spec understandable by XELB's el_client.el.  Systemtray, which is an XDG spec just like XSettings, defines the messages in the exwm code.  I'd say it belongs here.

minad commented on exwm-xsettings.el:
```diff
@@ -0,0 +1,343 @@
+
+  ;; Update the xsettings if/when the theme changes.
+  (add-hook 'enable-theme-functions #'exwm-xsettings--on-theme-change)
+  (add-hook 'disable-theme-functions #'exwm-xsettings--on-theme-change))
```
I suggest to make the update settings function public, such that it could be called from the user configuration where appropriate instead of registering these hooks. If desired the user can still register update-settings as theme hook.

EDIT: I saw that you also include code to switch between the light and dark variants of the Gtk theme. This is nice and also what I use in my configuration right now by reloading xsettingsd. Nevertheless it may be better to decouple the theme gluing code from the xsettings functionality here, and leave the coupling to the user configuration? I still hope to find configurable Gtk (and ideally also Qt) themes somewhere, which would allow me to transfer some of the basic colors from the Emacs theme (background, foreground), instead of only toggling between Adwaita Light and Dark. If such a thing would exist, the simple dark/light toggle wouldn't be enough.

Stebalien replied:
A user should be able to just call `(setopt exwm-xsettings-theme "NewTheme")`, they don't need to call these functions.

You're right that this code isn't, strictly speaking, necessary. Nor is it sufficient for all use-cases. But it probably covers _most_ cases and it's pretty convenient.

minad replied:
> A user should be able to just call (setopt exwm-xsettings-theme "NewTheme"), they don't need to call these functions.

Right. Exposing the update function is not necessary.

> You're right that this code isn't, strictly speaking, necessary. Nor is it sufficient for all use-cases. But it probably covers most cases and it's pretty convenient.

Yes, having this functionality here with the possibility to opt-out should not hurt. Maybe if the `exwm-xsettings-theme` variables are not set, the functionality could be completely disabled, no hooks installed?

Stebalien replied:
>  the functionality could be completely disabled, no hooks installed?

I considered it but instead made them no-ops when not required.

- The logic required for automatically adding-removing the hooks is more complex and the cost of running the hooks when they're "not used" is minimal.
- Adding them in `exwm-xsettings--init` gives the user a way to reliably remove them if they _really_ don't want this behavior.
- EXWM already hooks into `enable-theme-functions` and `disable-theme-functions` in a few places to detect changes in the background color (system tray and the background module).

minad replied:
Thanks! Makes sense. Looking forward to your PR getting merged.

minad commented on exwm-xsettings.el:
```diff
@@ -0,0 +1,343 @@
+  :initialize #'custom-initialize-default
+  :set (lambda (symbol value)
+         (set-default-toplevel-value symbol value)
+         (exwm-xsettings--update-settings)))
```
Extract this setter as a separate function and also use if for `exwm-xsettings-theme` and icon theme?

Stebalien replied:
Ah, yep. I missed that `(setop exwm-xsettings-theme ...)` doesn't actually apply immediately. Fixed.

Stebalien committed:
Trigger an XSETTINGS update when -theme and -icon-theme are set

medranocalvo commented:
Thank you @Stebalien.  I'll review as soon as possible.

minad commented:
As a data point, I've installed this `exwm-xsettings` package in my configuration about a week ago. It works flawlessly (theme switching and other settings). I think the only thing I am missing here is a little bit of documentation. Maybe an example configuration in the commentary would do?

This is what I use:

~~~ elisp
(require 'exwm-xsettings)
(setq exwm-xsettings-theme '("Adwaita" . "Adwaita-dark")
      exwm-xsettings '(("Xft/HintStyle" . "hintslight")
                       ("Xft/RGBA" . "rgb")
                       ("Xft/lcdfilter" . "lcddefault")
                       ("Xft/Antialias" . 1)
                       ("Xft/Hinting" . 1)
                       ("Xft/DPI" . 147456)))
(exwm-xsettings-enable)
~~~

Stebalien committed:
Add some documentation

Stebalien commented:
LGTY?

medranocalvo commented on exwm-xsettings.el:
```diff
@@ -0,0 +1,373 @@
+;;; exwm-xsettings.el --- XSETTINGS Module for EXWM -*- lexical-binding: t -*-
+
+;; Copyright (C) 2022 Free Software Foundation, Inc.
```
Is the year correct?  (No problem if it is.)

Stebalien replied:
I did write it in 2022, but updated it this year so I'll make it a range.

medranocalvo commented:
I have reviewed the code.  I answered a couple of questions.  I'm happy to merge when you think it's ready (e.g. remove TODO: comment).

This is also a great feature, thank you very much!

Stebalien committed:
Update copyright year for XSETTINGS

Stebalien committed:
Duplicate systemtray:-ClientMessage to xsettings:-ClientMessage

And fix the imports.

Stebalien commented:
Ok, I've addressed everything and it should be good to go now. Thanks for the review!

medranocalvo commented:
One final nit: up to now I've followed strictly the Emacs commit message conventions (https://github.com/ch11ng/exwm/wiki/Maintenance#commit-messages), mainly because @ch11ng did.  I use Magit and `C-x 4 a` and it is easy enough for me.  But maybe it's time to revise this "policy".  Should we follow Emacs commit message conventions?  @minad, I'd like you to weigh in as well.

minad commented:
In my projects I mostly don't follow the convention, so we could as well relax it. But if mostly people feel like keeping the convention I won't mind.

Stebalien commented:
So, there are two issues:

1. Commit messages.
2. Commits themselves.

With respect to the first, I kind of like Emacs's commit format as it tells you exactly what changed and why, but it's also kind of annoying. I don't have a strong preference either way (but I've updated the PR description with what the commit message _should_ be).

With respect to the second, the issue here is that GitHub doesn't have proper "patch sets". That leaves three options for addressing PR review.

1. Open a new PR (terrible).
4. Amend/rebase and push (difficult to review).
5. Keep adding new commits.

So I've gotten into the habit of doing the latter. Unfortunately, that means that the best way to merge PRs (without random commits) is, IMO, squash-merge.

---

For now, I'd:

1. Squash merge.
2. Keep the commit message format, but consider formatting the commit message (when squash-merging) on behalf of contributor if it's not too complex.

That is, unless the patch requires no changes and/or is a nicely formatted patch set. In that case, just rebase-merge.

medranocalvo commented:
@Stebalien I agree on both 1. and 2.

On 1., I find the consistent messages slightly helpful, it is kind of a semi-automatic summary.  The format allows the more usual commentary with the summary at the bottom (see git log for examples).  Using `C-x 4 a` from the Magit diff (it might work in VC?) makes the chore easy.  I've been editing most (?) contribution's commit messages and it's not been a big hassle for me.  I believe that we shouldn't require contributors to write it themselves.

On 2. I believe that most contributions have been single commits (not sure?).  I prefer the adding of further commits with the changes after review, but in the end it will be whatever the contributor does.  (We can follow your proposal between us.) Squashing and writing the final message seems a comfortable default.

minad commented on exwm-xsettings.el:
```diff
@@ -0,0 +1,384 @@
+
+(defcustom exwm-xsettings nil
+  "Custom XSETTINGS.
+These settings take precedence over `exwm-xsettings-theme' and `exwm-xsettings-icon-theme'."
```
@Stebalien We should break the long doc string lines here, since the byte compiler will otherwise complain. I have checkdoc and package-lint always enabled in my Elisp buffers which will warn me early on about such nits.

Stebalien replied:
Load-path issues were preventing it from linting the rest of the file.

Stebalien committed:
Wrap long documentation lines.

Stebalien commented on exwm-xsettings.el:
```diff
@@ -0,0 +1,386 @@
+  (pcase theme
+    ((cl-type string) theme)
+    (`(,(cl-type string) . ,(cl-type string))
+     (if (color-dark-p (color-name-to-rgb (face-background 'default)))
```
Ok, so, this requires emacs 28.1. Should I just not support this feature on older Emacs versions?

minad replied:
I wouldn't like if certain EXWM features are only available on newer Emacs versions. EXWM as a whole should work coherently on the supported Emacs version.

We could either implement the missing function for older Emacs versions or pull in Compat as dependency. Compat may be useful if we want to use other functionality. I cannot tell right now, but maybe this will be the case later (eg when we refactor some parts of the code).

Independent of this issue I suggest that we raise the dependency to 27.1 if there are no objections and support 27.1 for a few years. 27.1 is widely available on Linux distributions, including the ones with long time support.

minad replied:
See https://github.com/emacs-compat/compat/blob/dff3639b70f3c9916c5397056a51cdc3e60d5668/compat-28.el#L672-L699 for the relevant code in Compat.

Stebalien replied:
So, options are:

1. Copy some Emacs code here.
2. Depend on compat (may let us use newer emacs features elsewhere as well).
3. Remove the automatic dark mode switching (as pointed out by @minad earlier, users can handle dark-mode switching themselves).

At this point, I'm leaning towards option 3, actually. I _like_ the feature, but this is the second time we've tripped over it (and there's no reason we can't add it back in the future).

Preferences @minad, @medranocalvo?

minad replied:
My preference is 1, 2, 3 in this order. To elaborate:

1. This is the quickest solution in the short term, and easy to replace with Compat if we start using it.
2. In the long term I propose that we depend on Compat, since this will help us using newer APIs. This can be done as part of some refactoring. Compat will also be officially supported for Emacs core packages soon (see https://lists.gnu.org/archive/html/bug-gnu-emacs/2024-01/msg00660.html).
3. Removing the feature was my preference when I first reviewed this patch. But by now I am bit more pragmatic. I use the theme switching and it really seems the easiest way to enable this out of the box for the users. It would be unfortunate if users would have to reimplement this themselves.

minad replied:
@Stebalien How shall we proceed? I think it is okay to merge this into the current master branch to make it available to a wider user base. The addition is low risk since it must be enabled explicitly.

Stebalien replied:
Done. Sorry, I try to limit my Emacs hacking to the weekends (try being the operative word).

minad replied:
Okay, thanks! Since we cannot merge here, feel free to push this to https://github.com/emacs-exwm/exwm and close the PR here. I just installed the current version of exwm-xsettings.el in my configuration and gave it a final test.

Stebalien committed:
Implement a color-dark-p function on Emacs versions < 29

Stebalien commented on exwm-xsettings.el:
```diff
@@ -181,6 +181,21 @@ These settings take precedence over `exwm-xsettings-theme' and
+              "Whether RGB is more readable against white than black."
+              (unless (<= 0 (apply #'min rgb) (apply #'max rgb) 1)
+                (error "RGB components %S not in [0,1]" rgb))
+              (let* ((r (expt (nth 0 rgb) 2.2))
+                     (g (expt (nth 1 rgb) 2.2))
+                     (b (expt (nth 2 rgb) 2.2))
+                     (y (+ (* r 0.2126) (* g 0.7152) (* b 0.0722))))
+                (< y 0.325)))
```
I shrunk this code down a bit and inlined the constant (we're not maintaining this code here, so readability isn't a goal.

minad replied:
Inlining the constant makes sense.

minad commented on exwm-xsettings.el:
```diff
@@ -181,6 +181,21 @@ These settings take precedence over `exwm-xsettings-theme' and
 
+(eval-and-compile
+  (fset 'exwm-xsettings--color-dark-p
+        (if (< emacs-major-version 29)
```
I suggest to write this in a slightly different form with `defalias`, which I consider cleaner:

~~~ elisp
(defalias 'exwm-xsettings--color-dark-p 
   (if (eval-when-compile (< emacs-major-version 29)) ...))
~~~

Stebalien replied:
Fixed and amended.

Stebalien commented:
Moved to https://github.com/emacs-exwm/exwm/pull/2. I kept it on my local fork for now, although I'll probably start pushing branches directly to emacs-exwm/exwm in the future.
